x-pack/filebeat/input/awss3/sqs_acker.go

```diff
@@ -0,0 +1,212 @@
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+)
+
+var ackerIDCounter *atomic.Uint64
+
+func init() {
+	ackerIDCounter = atomic.NewUint64(0)
+}
+
+// EventACKTracker tracks the publishing state of S3 objects. Specifically
+// it tracks the number of message acknowledgements that are pending from the
+// output. It can be used to wait until all ACKs have been received for one or
+// more S3 objects.
+type EventACKTracker struct {
+	ID uint64
+
+	EventsAcked       *atomic.Uint64
+	EventsDropped     *atomic.Uint64
+	EventsPublished   *atomic.Uint64
+	EventsToBeTracked *atomic.Uint64
+
+	ctx        context.Context
+	cancel     context.CancelFunc
+	deletionWg *sync.WaitGroup
+
+	mutex *sync.Mutex
+
+	msg             *types.Message
+	ReceiveCount    int
+	start           time.Time
+	processingErr   error
+	Handles         []s3ObjectHandler
+	keepaliveCancel context.CancelFunc
+	keepaliveWg     *sync.WaitGroup
+	msgHandler      sqsProcessor
+	log             *logp.Logger
+}
+
+func NewEventACKTracker(ctx context.Context, deletionWg *sync.WaitGroup) *EventACKTracker {
+	ctx, cancel := context.WithCancel(ctx)
+	acker := &EventACKTracker{
+		ID:                ackerIDCounter.Inc(),
+		ctx:               ctx,
+		cancel:            cancel,
+		deletionWg:        deletionWg,
+		mutex:             new(sync.Mutex),
+		EventsAcked:       atomic.NewUint64(0),
+		EventsDropped:     atomic.NewUint64(0),
+		EventsPublished:   atomic.NewUint64(0),
+		EventsToBeTracked: atomic.NewUint64(0),
+	}
+
+	go func() {
+		t := time.NewTicker(500 * time.Microsecond)
+		defer t.Stop()
+
+		for {
+			<-t.C
+			if !acker.FullyTracked() {
+				continue
+			}
+
+			acker.cancelAndFlush()
+			return
+		}
+	}()
+
+	return acker
+}
+
+func (a *EventACKTracker) cancelAndFlush() {
+	a.cancel()
+	a.FlushForSQS()
+}
+
+// MarkSQSProcessedWithData Every call after the first one is a no-op
+func (a *EventACKTracker) MarkSQSProcessedWithData(msg *types.Message, publishedEvent uint64, receiveCount int, start time.Time, processingErr error, handles []s3ObjectHandler, keepaliveCancel context.CancelFunc, keepaliveWg *sync.WaitGroup, msgHandler sqsProcessor, log *logp.Logger) {
+	// We want to execute the logic of this call only once, when the ack mutex was locked on init
+	if a.EventsToBeTracked.Load() > 0 {
+		return
+	}
+
+	a.msg = msg
+	a.EventsToBeTracked = atomic.NewUint64(publishedEvent)
+	a.ReceiveCount = receiveCount
+	a.start = start
+	a.processingErr = processingErr
+	a.Handles = handles
+	a.keepaliveCancel = keepaliveCancel
+	a.keepaliveWg = keepaliveWg
+	a.msgHandler = msgHandler
+	a.log = log
+}
+
+func (a *EventACKTracker) FullyTracked() bool {
+	eventsToBeTracked := a.EventsToBeTracked.Load()
+	if eventsToBeTracked == 0 {
+		return false
+	}
+
+	// This is eating its own tail: we should check for dropped+published, but then we won't wait for acked.
+	// Acked might not be equal to published?
+	return a.EventsDropped.Load()+a.EventsAcked.Load() == eventsToBeTracked
```

Comment:
`EventsDropped` shouldn't be included here, and probably shouldn't be tracked at all since it has no effect that the input can act on (see comments on `(*eventListener).AddEvent` and `FlushForSQS`)

Reply:
I see we still need to track.
We know that a sqs message has 10 events
3 of them were dropped
7 acked

We sum the two above that's 7+3 == 10, so we know every events in the sqs message was taken care by the queue and we can delete the message without waiting anything else.